### ISSUE TITLE
All chart types imported / Can't let Webpack load CSS

### ISSUE THREAD

**Issue by dhardtke** [swimlane/ngx-charts#623]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here
```

**Current behavior**
When I import the `LineChartModule` for example from `@swimlane/ngx-charts` webpack ends up bundling the complete **index.js** from `@swimlane/ngx-charts/release`. Instead, I only want to import and actually have the chart types bundled that I am actually using, so I do this instead:

`import { LineChartModule } from "@swimlane/ngx-charts/release/line-chart/line-chart.module";`

This fixes the problem with all chart types being bundled, but instead I now get several CSS loading errors, even if I load the files manually in my `vendor.ts` chunk:

> GET http://localhost:8081/tooltip.component.css 404 (Not Found)
> app.js:1623 ERROR Error: Uncaught (in promise): Failed to load tooltip.component.css

**Expected behavior**
I should be able to load CSS "manually", i.e. let the bundler (webpack in my case) handle loading CSS.

One fix would be to ship es2015 bundles as well, which would probably allow me to use `@swimlane/ngx-charts` in my imports directly.

**What is the motivation / use case for changing the behavior?**
I don't want to waste bandwidth and load bloated JS on the client side.

* **ngx-charts version:** x.x.x
6.1.0

* **Angular version:** 2.x.x
5.0.0

* **Browser:** Chrome, Firefox (both latest)

* **Language:** TypeScript 2.4.2 with ES2015 as target

**Comment by bodinaren** [comment]
This is a significant problem for us too. The extra bandwidth is just too unreasonable.

Is it as easy as adding the `src` directory to the includes files in `package.json`?
```
"files": [
  "release",
  "src"
]
```
You will then be able to use this:
```
import { LineChartModule } from "@swimlane/ngx-charts/src/line-chart/line-chart.module";
```
Or is there a reason you would not want this?

**Comment by marjan-georgiev** [comment]
Importing directly from the source could cause problems on projects that do not use typescript, as those are not transpiled.

Curious, doesn't webpack tree shaking get rid of the code you are not using during your bundling process?

**Comment by dhardtke** [comment]
No, it does not (unfortunately).

Either because Webpack's Tree Shaking is broken, or because of a configuration problem on our end (which I doubt).

I agree that is not the best way to do things, but I don't see another option for only including the actual chart types that we are using in our bundled output.

**Comment by bodinaren** [comment]
Tree shaking, unfortunately, only looks at the imports. Since the index.js imports every module that means every module is bundled.

**Comment by bodinaren** [comment]
I was just now trying to update to Angular 5. The approach I suggested causes problems because with the default settings the .ts files is not being compiled (see https://github.com/angular/angular-cli/issues/8284#issuecomment-343673297). Doing as suggested (adding `node_modules/@swimlane/ngx-charts/src` to `includes` in tsconfig.json gives error of missing typings for jasmine when it tried to compile `spec.ts` files. This could of course also be fixed (I stopped here however) but this is clearly not the optimal approach (https://github.com/angular/angular-cli/issues/8284#issuecomment-343713685).

I saw that AotPlugin is not used https://github.com/swimlane/ngx-charts/blob/master/config/webpack.package.js#L79-L83, perhaps that's something that could fix this?

**Comment by dhardtke** [comment]
It all comes down to your webpack config.

Are you using the angular cli? Because we are not, we are using a custom webpack configuration with Angular's `AngularCompilerPlugin`.

Notice: The `AotPlugin` was renamed to the `AngularCompilerPlugin` for Angular 5.

**Comment by bodinaren** [comment]
Yeah, I'm using the angular cli.

I'm really not very good at this cli/webpack thins, but, is there no way
you can just make webpack/AngularCompilerPlugin compile the CSS into the
components instead with `styleUrls` of keeping the CSS file separate
`styles`? This, to me, seems to be the problem.

Den fre 17 nov. 2017 20:57Dominik Hardtke <notifications@github.com> skrev:

**Comment by bodinaren** [comment]
Yes, I am using the angular cli. I think most people who use ngx-charts is?

I've been looking around and reading a bit. As I understand it, the crux of the matter is the styleUrls. What we need is for the styles to be inlined instead. Most of what I find references this library [ludohenin/gulp-inline-ng2-template](https://github.com/ludohenin/gulp-inline-ng2-template). But is this perhaps something you could consider implementing into your build process?

I would try it out myself but to be honest, the details of how to use this is beyond me though, I'm definitely not well versed in the webpack/gulp world.

**Comment by dhardtke** [comment]
Also, I think we might be running into https://github.com/webpack/webpack/issues/2867 which is why I do not import all modules from "NgxChartsModule" but rather from each chart's individual module itself.

**Comment by dhardtke** [comment]
Fixed with #1004
